WireDomain/Sources/WireDomain/Synchronization/IncrementalSync.swift

```
@@ -75,7 +81,24 @@ public struct IncrementalSync: IncrementalSyncProtocol {
         } catch {
             logger.debug("incremental sync interrupted, tearing down...")
             await pushChannel.close()
-            throw error
+
+            switch error {
+            case let apiError as UpdateEventsAPIError:
+                switch apiError {
+                case .notFound, .invalidParameters:
+                    logger.debug("too many missed events.. recovering")
+                    // nullifying the last event ID since we missed events and we want to
+                    // reset with a full sync (initial + incremental)
+                    updateEventsStore.storeLastEventID(id: nil)
+                    try await messageStore.addPotentialGapSystemMessage()
+                    onMissedEvents()
```

suggestion: I think it would be better that the incremental sync throws an error like `missingEvents`, which is handled by the `SyncAgent` that would then trigger the initial sync. For me the `IncrementalSync` should just try to do what it's ask to do and report any failures, where as the `SyncAgent` is responsible for starting and stopping various syncs. What do you think?

---

that was my initial thought as well but then I figured in that case the SyncAgent would not only be responsible of high level sync operations (starting / stopping syncs) but also add potential gap system message and clean up the last event ID which I think is outside of its responsibilities because way too specific. 

In my opinion, we should report to the SyncAgent only for sync operations and keep the more specific logic closer to the object its related to  (i.e add potential system message, clean up the last event ID).

---

@johnxnguyen For what is worth, on async stream I don't throw an error (since it would close the websocket and avoid acknowledgement of event) instead I use a delegate to inform the SyncAgent to do the initialSync then I can continue on my flow.

---

I think the incremental can add the system message and then throw the error so the sync agent can perform the initial sync. Doing so should update the last event id already as a first step so there's no cleanup really necessary, right?

@netbe I would say that closing the push channel to perform the initial sync and then reopen the push channel is the behavior we want. Or is there any issue in doing this? Maybe one I can think of is that it may create a window of time where events are pushed to the NSE.

---

For me, the incremental sync can still clear the last event id and add the system message, but I think it should throw the error and let the `SyncAgent` handle it by triggering the appropriate sync.

---

so maybe better not to close the pushChannel and wait until initial sync is done

---

thanks for the feedbacks, I'll close this PR and open a new one and do the changes from 4.0 as there's too much discrepancies between branches now.